brave/src/main/java/brave/internal/handler/SafeSpanListener.java

```diff
@@ -0,0 +1,165 @@
+  // Array ensures no iterators are created at runtime
+  public static SpanListener create(SpanListener[] handlers) {
+    if (handlers.length == 0) return SpanListener.NOOP;
+    if (handlers.length == 1) return new Single(handlers[0]);
+    return new Multiple(handlers);
+  }
+
+  @Override
+  public void onCreate(@Nullable TraceContext parent, TraceContext context, MutableSpan span) {
+    try {
+      doOnCreate(parent, context, span);
+    } catch (Throwable t) {
+      propagateIfFatal(t);
+      Platform.get().log("error handling create {0}", context, t);
+    }
+  }
+
+  @Override public void onAbandon(TraceContext context, MutableSpan span) {
+    try {
+      doOnAbandon(context, span);
+    } catch (Throwable t) {
+      propagateIfFatal(t);
+      Platform.get().log("error handling abandon {0}", context, t);
+    }
+  }
+
+  @Override public void onFlush(TraceContext context, MutableSpan span) {
+    try {
+      doOnFlush(context, span);
+    } catch (Throwable t) {
+      propagateIfFatal(t);
+      Platform.get().log("error handling flush {0}", context, t);
+    }
+  }
+
+  @Override public void onOrphan(TraceContext context, MutableSpan span) {
+    try {
+      doOnOrphan(context, span);
+    } catch (Throwable t) {
+      propagateIfFatal(t);
+      Platform.get().log("error handling orphan {0}", context, t);
+    }
+  }
+
+  @Override public void onFinish(TraceContext context, MutableSpan span) {
+    try {
+      doOnFinish(context, span);
+    } catch (Throwable t) {
+      propagateIfFatal(t);
+      Platform.get().log("error handling finish {0}", context, t);
+    }
+  }
+
+  abstract void doOnCreate(@Nullable TraceContext parent, TraceContext context, MutableSpan span);
+
+  abstract void doOnAbandon(TraceContext context, MutableSpan span);
+
+  abstract void doOnFlush(TraceContext context, MutableSpan span);
+
+  abstract void doOnOrphan(TraceContext context, MutableSpan span);
+
+  abstract void doOnFinish(TraceContext context, MutableSpan span);
+
+  static final class Single extends SafeSpanListener {
+    final SpanListener delegate;
+
+    Single(SpanListener delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    void doOnCreate(@Nullable TraceContext parent, TraceContext context, MutableSpan span) {
+      delegate.onCreate(parent, context, span);
+    }
+
+    @Override void doOnAbandon(TraceContext context, MutableSpan span) {
+      delegate.onAbandon(context, span);
+    }
+
+    @Override void doOnFlush(TraceContext context, MutableSpan span) {
+      delegate.onFlush(context, span);
+    }
+
+    @Override void doOnOrphan(TraceContext context, MutableSpan span) {
+      delegate.onOrphan(context, span);
+    }
+
+    @Override void doOnFinish(TraceContext context, MutableSpan span) {
+      delegate.onFinish(context, span);
+    }
+
+    @Override public String toString() {
+      return delegate.toString();
+    }
+  }
+
+  static final class Multiple extends SafeSpanListener {
```

Comment:
Not sure if it's necessarily better, but I was sort of expecting there to just be one implementation of `SafeSpanListener` and this would actually be a `SpanListener`. We could possibly make it public too since it could be helpful for users to be able to combine multiple listeners but don't have to right now.

Reply:
I see what you mean.. good idea on only one safe thing.. I can go back and address the others.

I don't think we're ready to support user-side compositing as it is unlikely they will do a better job than we would. Probably better for us to just say "don't composite on your own" then wait for a scream? (this would apply to the several other places we do similar compositing). Not doing user-level compositing also keeps the api sprawl down.. brave's already collecting quite a number of public types...

Reply:
Not exposing a compositing listener to public unless we find a use for it in the future sgtm